### DISCUSSION TITLE
Missing ES6 features: How Mirth tackles discontinuation of Rhino?

### DISCUSSION THREAD

**Discussion by tom08zehn** [nextgenhealthcare/connect#4508] [Q&A]
I'm super happy with Mirth and was very excited when I updated to the latest version in order to get also latest Rhino with ES6 features. Then I noticed all the limitations for ES6 and was very disappointed. Even some of the simplest ES6 features are not supported: https://mozilla.github.io/rhino/compat/engines.html

This is not the "fault" of Mirth but as Rhino development is discontinued I wonder how Mirth tackles the situation (and user happiness) running a very outdated JavaScript engine?

**Comment by brentmoen** [comment]
There is an open issue (#4409) on this topic. To summarize, it will be necessary to migrate from Rhino at some point. It may be a while before that happens though since it will be a costly effort for a couple of reasons:

1. Rhino is very tightly coupled with the Mirth Connect messaging engine
2. The E4X extension is heavily relied upon and is not supported by modern JS engines

A likely solution will involve making the JS engine selectable, with Rhino being the default and V8 (or something else) being an option. Channels that use E4X would not be compatible with the newer engine.
